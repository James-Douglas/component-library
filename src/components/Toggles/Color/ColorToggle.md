# ColorToggle component
The CTM ColorToggle component. Renders a color-selection toggle.
## Usage
~~~js
<ColorToggle
  id={id}
  label={label}
  backgroundColor={backgroundColor}
  fontColor={fontColor}
  value={value}
  selectedId={selectedId}
  name={name}
  invalid={invalid}
  disabled={disabled}
  autofill={autofill}
  onToggle={onToggle}
/>
~~~

## Props
The ToggleGroup component accepts the following props:

1. `id` *(required) string*
    * Unique identifier for the toggle
    
2. `label` *(required) string*
    * Label for the toggle.
    
3. `backgroundColor` *(required) string*
    * The background color to be displayed (the color to be selected).
    
4.  `fontColor` *(optional) string*
    * Font color to be applied to the label (default: 'black')

5.  `value` *(optional) string*
    * The value to be applied to the input field.
    
6.  `selectedId` *(optional) string*
    * The id of the currently selected (toggled on) toggle.
    
7.  `name` *(optional) string*
    * The name to be applied to the input field.
    
8.  `invalid` *(optional) boolean*
    * Applies invalid styling when true.
    
9.  `disabled` *(optional) boolean*
    * Disables the toggle when true.

10.  `autofill` *(optional) boolean*
    * Autofills (toggles) when true.
    
11. `onToggle` *(optional) function*
    * Handler function called when a toggle is toggled on (called with the id of the toggle).
