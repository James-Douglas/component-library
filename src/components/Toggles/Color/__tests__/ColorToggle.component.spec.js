import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ColorToggle, { getDisplayBackgroundColor, getAnimationStyle, getDisplayLabel } from '../ColorToggle.component';

describe('getDisplayBackgroundColor()', () => {
  it('returns background color for non-white', () => {
    expect(getDisplayBackgroundColor('black')).toEqual('black');
  });

  it('returns ivory for white background', () => {
    expect(getDisplayBackgroundColor('white')).toEqual('ivory');
  });
});

describe('getAnimationStyle', () => {
  it('returns correct style for un-selected toggle', () => {
    expect(getAnimationStyle('test-a', 'test-b', 'yellow')).toEqual({ backgroundColor: 'yellow' });
  });

  it('returns correct style for selected toggle', () => {
    const result = getAnimationStyle('test-a', 'test-a', 'yellow');
    expect(result).toEqual({ backgroundColor: 'yellow', height: '100%' });
  });
});

describe('getDisplayLabel', () => {
  it('returns label when label provided', () => {
    expect(getDisplayLabel('test label', null)).toEqual('test label');
  });

  it('returns correct color label when no label provided (non-white)', () => {
    expect(getDisplayLabel(null, 'black')).toEqual('Black');
  });

  it('returns correct color label when no label provided (white)', () => {
    expect(getDisplayLabel(null, 'ivory')).toEqual('White');
  });
});

describe('ColorToggle', () => {
  it('renders with minimal props', () => {
    const { getByText, container } = render(<ColorToggle backgroundColor="black" id="test-a" />);
    expect(container.innerHTML).toMatchSnapshot();
    expect(getByText('Black')).toBeInTheDocument();
    const wrapper = container.querySelector('.scoped-toggle');
    expect(wrapper).toHaveClass('black');
    const input = container.querySelector('input');
    expect(input).toHaveAttribute('id', 'test-a');
    expect(input).toHaveAttribute('tabIndex', '0');
  });

  it('applies fontColor when provided', () => {
    const { container } = render(<ColorToggle backgroundColor="black" id="test-a" fontColor="white" />);
    expect(container.querySelector('.scoped-toggle')).toHaveClass('white');
  });

  it('calls onToggle when change event fires', () => {
    const onToggleCb = jest.fn();
    const { container } = render(<ColorToggle backgroundColor="black" id="test-a" fontColor="white" onToggle={onToggleCb} />);
    const element = container.querySelector('.toggle');
    fireEvent.click(element);
    expect(onToggleCb).toHaveBeenCalled();
    expect(onToggleCb.mock.calls[0][0]).toEqual({ id: 'test-a', value: '' });
  });
});
