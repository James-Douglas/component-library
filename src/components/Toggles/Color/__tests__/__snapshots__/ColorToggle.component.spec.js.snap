// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorToggle renders with minimal props 1`] = `
"<div class=\\"scoped-toggle black\\"><style></style><span class=\\"flex toggle\\"><style></style><input tabindex=\\"0\\" class=\\"
          toggle-input 
          
          
        \\" id=\\"test-a\\" type=\\"radio\\" name=\\"\\" value=\\"\\"><style></style><label class=\\"toggle-label transition\\" for=\\"test-a\\"><span class=\\"colour-toggle\\"><span class=\\"content\\">Black</span><span class=\\"border-colour\\" style=\\"background-color: black;\\"></span></span></label></span></div>"
`;
